pkg/graveler/ref/manager.go

```diff
@@ -412,19 +445,78 @@ func (m *Manager) ListCommits(ctx context.Context, repositoryID graveler.Reposit
 }
 
 func (m *Manager) FillGenerations(ctx context.Context, repositoryID graveler.RepositoryID) error {
-	_, err := m.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		return tx.Exec(`WITH RECURSIVE cte AS
-								(
-									SELECT id, 1 AS generation,creation_date
-									FROM graveler_commits
-									WHERE repository_id = $1 AND parents IS NULL OR array_length(parents,1) = 0
-									UNION ALL
-									SELECT DISTINCT(c.id), cte.generation+1 AS generation,c.creation_date
-									FROM graveler_commits c INNER JOIN cte
-									ON cte.id = ANY(c.parents) WHERE repository_id = $1
-								)
-								UPDATE graveler_commits u SET generation = cte.generation FROM cte WHERE u.id = cte.id;`,
-			repositoryID)
+	// update commitNodes' generation in nodes "tree" using BFS algorithm.
+	// using a queue implementation
+	// adding a node to the queue only after all of its parents were visited in order to avoid redundant visits of nodesCommitsIDs
+	nodes, err := m.createCommitsIDsMap(ctx, repositoryID)
+	if err != nil {
+		return err
+	}
+	rootsCommitIDs := m.getRootNodes(nodes)
+	m.mapCommitNodesToChildren(nodes)
+	m.addGenerationToNodes(nodes, rootsCommitIDs)
+	_, err = m.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
+		return nil, m.updateCommitGeneration(tx, repositoryID, nodes)
 	})
 	return err
 }
+
+func (m *Manager) createCommitsIDsMap(ctx context.Context, repositoryID graveler.RepositoryID) (map[graveler.CommitID]*CommitNode, error) {
+	iter, err := m.ListCommits(ctx, repositoryID)
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Close()
+
+	nodes := make(map[graveler.CommitID]*CommitNode)
+	for iter.Next() {
+		commit := iter.Value()
+		parentsToVisit := map[graveler.CommitID]struct{}{}
+		for _, parentID := range commit.Parents {
+			parentsToVisit[parentID] = struct{}{}
+		}
+
+		nodes[commit.CommitID] = &CommitNode{parentsToVisit: parentsToVisit}
+	}
+	if err := iter.Err(); err != nil {
+		return nil, fmt.Errorf("getting value from iterator: %w", err)
+	}
+	return nodes, nil
+}
+
+func (m *Manager) getRootNodes(nodes map[graveler.CommitID]*CommitNode) []graveler.CommitID {
+	var rootsCommitIDs []graveler.CommitID
+	for commitID, commit := range nodes {
+		if len(commit.parentsToVisit) == 0 {
+			rootsCommitIDs = append(rootsCommitIDs, commitID)
+		}
+	}
+	return rootsCommitIDs
+}
+
+func (m *Manager) mapCommitNodesToChildren(nodes map[graveler.CommitID]*CommitNode) {
+	for commitID, commitNode := range nodes {
+		// adding current node as a child to all parents in commitNode.parentsToVisit
+		for parentID := range commitNode.parentsToVisit {
+			nodes[parentID].children = append(nodes[parentID].children, commitID)
+		}
+	}
+}
+
+func (m *Manager) addGenerationToNodes(nodes map[graveler.CommitID]*CommitNode, rootCommits []graveler.CommitID) {
```

Comment:
I think this function should be tested.  Could open a TODO/issue to do this rather than delay this PR.